x-pack/plugin/watcher/src/main/java/org/elasticsearch/xpack/watcher/notification/WebhookService.java

```diff
@@ -121,41 +122,80 @@ public Action.Result execute(
+        HttpRequest originalRequest = action.getRequest().render(templateEngine, model);
 
-        // If applicable, add the extra token to the headers
-        boolean tokenAdded = false;
-        WebhookAccount account = getAccount(NAME);
-        if (this.additionalTokenEnabled && account.hostTokenMap.size() > 0) {
-            // Generate a string like example.com:9200 to match against the list of hosts where the
-            // additional token should be provided. The token will only be added to the headers if
-            // the request matches the list.
-            String reqHostAndPort = request.host() + ":" + request.port();
-            if (Strings.hasText(account.hostTokenMap.get(reqHostAndPort))) {
-                // Add the additional token
-                tokenAdded = true;
-                request = request.copy().setHeader(TOKEN_HEADER_NAME, account.hostTokenMap.get(reqHostAndPort)).build();
-            }
+        if (ctx.simulateAction(actionId)) {
+            HttpRequest request = maybeModifyHttpResponse(originalRequest);
+            // If the request was modified, then the request has had the token added
+            boolean tokenAdded = originalRequest != request;
+            // Skip execution, return only the simulated (and redacted if necessary) response
+            return new WebhookAction.Result.Simulated(
+                tokenAdded ? request.copy().setHeader(TOKEN_HEADER_NAME, WatcherXContentParser.REDACTED_PASSWORD).build() : request
+            );
         }
 
+        final Tuple<HttpRequest, HttpResponse> respTup = modifyAndExecuteHttpRequest(originalRequest);
+        final HttpRequest request = respTup.v1();
+        final HttpResponse response = respTup.v2();
+        // If the request was modified, then the request has had the token added
+        final boolean tokenAdded = originalRequest != request;
+
         final Function<HttpRequest, HttpRequest> redactToken = tokenAdded
             ? req -> req.copy().setHeader(TOKEN_HEADER_NAME, WatcherXContentParser.REDACTED_PASSWORD).build()
             : Function.identity();
 
-        if (ctx.simulateAction(actionId)) {
-            // Skip execution, return only the simulated (and redacted if necessary) response
-            return new WebhookAction.Result.Simulated(redactToken.apply(request));
-        }
-
-        HttpResponse response = httpClient.execute(request);
-
         if (response.status() >= 400) {
             return new WebhookAction.Result.Failure(redactToken.apply(request), response);
         } else {
             return new WebhookAction.Result.Success(redactToken.apply(request), response);
         }
     }
 
+    /**
+     * Makes any additional modifications to the {@link HttpRequest} if necessary.
+     * If no modifications are made the same instance is returned, otherwise a new
+     * HttpRequest is returned.
+     */
+    private HttpRequest maybeModifyHttpResponse(HttpRequest request) {
```

Comment:
This ought to be `maybeModifyHttpRequest` right? It's not modifying the response.

Reply:
Yep, that's a typo, I pushed a commit to fix this.